### PR TITLE
 git-worktree-switcher: init at 0.2.4

### PR DESCRIPTION
**Pull request by jiriks74**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jiriks74** [comment]
I think that all everything should now work as expected. I'll look into doing tests as that's something I am not familiar with.
But I think that the rest can be reviewed.

**Comment by mateusauler** [comment]
Awesome! I'm willing to co-maintain the package, if you'd like.

**Comment by jiriks74** [comment]
> Awesome! I'm willing to co-maintain the package, if you'd like.

I'll gladly add you if you're in the maintainers file.

**Comment by jiriks74** [comment]
@mateusauler

I didn't find you in [the maintainers file](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix).
I can add you if you fork my changes, follow [the README](https://github.com/NixOS/nixpkgs/tree/master/maintainers#individual-maintainer-list) for adding maintainers and make a PR. I **cannot** add you as when you add a GPG signature to a maintainer the commit **has to** be signed with the key added.

**Comment by jiriks74** [comment]
I just added this package and module to my NixOS config. I had to make a modification to the module though:

```diff
diff --git a/modules/nixos/git-worktree-switcher.nix b/modules/nixos/git-worktree-switcher.nix
index f1462ab..490c3a2 100644
--- a/modules/nixos/git-worktree-switcher.nix
+++ b/modules/nixos/git-worktree-switcher.nix
@@ -2,6 +2,7 @@
   config,
   pkgs,
   lib,
+  ...
 }:
 
 let
```

or I'd get

```nix
       error: function 'anonymous lambda' called with unexpected argument 'outputs'
       at /nix/store/qc8rd1yhqd06yjfcnpigkfpnfgmyz2dy-source/modules/nixos/git-worktree-switcher.nix:1:1:
            1| {
             | ^
            2|   config,
```

So it seems like the missing `...` creates an issue with flakes.

---

Is the no `...` requirement only for packages or for modules too? (I understood from the docs to avoid it if possible)

I'll be looking through other modules and adding it back if I see `...` being commonly used.

**Comment by mateusauler** [comment]
Whoops, forgot to format the maintainers section in the package file.

**Comment by jiriks74** [comment]
> Whoops, forgot to format the maintainers section in the package file.

I'll take care of it. I'll also add back the `...` since I misunderstood the docs and it's kinda needed for the module to work.

**Comment by mateusauler** [comment]
> I'll take care of it.

Thanks.

> I'll also add back the `...` since I misunderstood the docs and it's kinda needed for the module to work.

I'll test it on my config as well and report any problems other than this one.

**Comment by FliegendeWurst** [comment]
Please make sure your commit messages follow the guidelines.

**Comment by mateusauler** [comment]
I believe it would be best not to include github's PR merge commit. As far as I can tell, just running `git rebase -i HEAD~2` and dropping it doesn't break anything.

**Comment by jiriks74** [comment]
> I believe it would be best not to include github's PR merge commit. As far as I can tell, just running git rebase -i HEAD~2 and dropping it doesn't break anything.

Makes sense. Should've thought of that before XD.
I droped the last two commits (the merge commit didn't show in rebase) and then I merge it again using git itself.

Hopefully everything's ready now.

**Comment by mateusauler** [comment]
I noticed the module isn't in `nixos/modules/module-list.nix` and doesn't work in my config probably because of that.

**Comment by jiriks74** [comment]
> I noticed the module isn't in nixos/modules/module-list.nix and doesn't work in my config probably because of that.

Should be fixed now.

**Comment by jiriks74** [comment]
Well now I can either `git rebase` and break @mateusauler's commit again or I can `git merge` and have the merge commit there.

Which one you think would be better for this situation?

**Comment by jiriks74** [comment]
Or I can remove @mateusauler's commit altogether, finish this, and add him as a comaintainer once everything is done.

**Comment by mateusauler** [comment]
Feel free to break my commit or remove it :). If you end up removing it, I'll create a PR, after this one is merged, and add myself back.
I'll also test everything later today and report back here.

**Comment by jiriks74** [comment]
> After all's working we can add it back or I'll create a PR to add myself as a maintainer.

I think that this is the best course of action too. I'll be looking out for your PR.

I'll look into testing this probably this evening. I have all the changes in my private Flake for my systems so the code itself works, it only needs to be added `nixpkgs` in a way that works.

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (pluiedev) [inline nixos/modules/programs/git-worktree-switcher.nix:12]: This seems awfully redundant. I think we can just have one init script with the shell name substituted in

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (pluiedev) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:11]: Consider using finalAttrs

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (pluiedev) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:53]: Avoid `with lib;`

**Review by pluiedev — APPROVED** [review]
Overall LGTM, only some minor nitpicks - thanks for contributing to Nixpkgs!

File: maintainers/maintainer-list.nix
Comment (pluiedev) [inline maintainers/maintainer-list.nix:10360]: Oh and this should be a separate commit

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (jiriks74) [reply pkgs/by-name/gi/git-worktree-switcher/package.nix:11]: I'll look into this. I've basically modified the `thefuck` package to get this as I am not 100% familiar with all the options.

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:58]:
```suggestion
  meta = {
    homepage = "https://github.com/mateusauler/git-worktree-switcher";
    description = "Switch between git worktrees with speed.";
    license = lib.licenses.mit;
    platforms = lib.platforms.all;
    maintainers = with lib.maintainers; [ jiriks74 ];
```

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (pluiedev) [reply pkgs/by-name/gi/git-worktree-switcher/package.nix:53]: > What should I use instead then? I haven't found much documentation about this.

See FliegendeWurst's proposed changes

**Comment by jiriks74** [comment]
I've incorporated all the changes. The package builds as of now. I'll test the functionality shortly.

**Comment by jiriks74** [comment]
I just tested the package with `nixpkgs` set to my local repo in my OS config. Worked without issues.

File: maintainers/maintainer-list.nix
Comment (jiriks74) [reply maintainers/maintainer-list.nix:10360]: Fixed in b489ea0f4e1904db70c37bfbfe481fa2cc84c9f3

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (jiriks74) [reply pkgs/by-name/gi/git-worktree-switcher/package.nix]: Fixed in 7c2dda4abd2b5cef40276badf4f57dd391bc209f

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (jiriks74) [reply pkgs/by-name/gi/git-worktree-switcher/package.nix]: Fixed in 7c2dda4abd2b5cef40276badf4f57dd391bc209f

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (jiriks74) [reply nixos/modules/programs/git-worktree-switcher.nix]: Fixed in 3d30c8d9340903897037bdf0d8f4eed08e56fa12

**Comment by mateusauler** [comment]
I've also tested it and everything seems fine. I found a bug with the script itself, which I'll fix later.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (xanderio) [inline nixos/doc/manual/release-notes/rl-2411.section.md:204]: 24.11 has been release by now, could you please move this to the 25.05 release notes

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (xanderio) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:34]: I don't think we should add functionality to a program via a patch in nixpkgs, could you send this upstream?

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:27]:
```suggestion
    programs.zsh.interactiveShellInit = lib.optionalString config.progams.zsh.enable (initScript "zsh");
```

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:13]:
```suggestion
  initScript =
    shell:
    if (shell != "fish") then
      ''
        eval $(${getExe pkgs.git-worktree-switcher} init ${shell})
      ''
    else
      ''
        ${getExe pkgs.git-worktree-switcher} init ${shell} | source
      '';```

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:26]:
```suggestion
    programs.bash.interactiveShellInit = initScript "bash";
```

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:28]:
```suggestion
    programs.fish.interactiveShellInit = lib.optionalString config.programs.fish.enable (initScript "fish");
```

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:9]:
```suggestion

```

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (xanderio) [inline nixos/modules/programs/git-worktree-switcher.nix:10]:
```suggestion
  cfg = config.programs.git-worktree-switcher;
```

**Review by xanderio — CHANGES_REQUESTED** [review]
Thanks for your contribution!

The major thing that doesn't look right to me is the fish config part. I've suggested some changes that should make it work. Could you please tests them :)

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (jiriks74) [reply pkgs/by-name/gi/git-worktree-switcher/package.nix:34]: We already started a discussion on this. I agree that it shouldn't be here.

File: nixos/modules/programs/git-worktree-switcher.nix
Comment (jiriks74) [reply nixos/modules/programs/git-worktree-switcher.nix:13]: May I ask why would you want use 2 different scripts when we can just use `eval $(${getExe pkgs.git-worktree-switcher} init ${shell})` for everything?

Comment (xanderio) [reply nixos/modules/programs/git-worktree-switcher.nix]: As fish is not a POSIX shell [`eval`](https://fishshell.com/docs/current/cmds/eval.html) isn't the intended way to load code into the shell, while `eval` may work in this case, let's please stick to piping into [`source`](https://fishshell.com/docs/current/cmds/source.html).

Comment (jiriks74) [reply nixos/modules/programs/git-worktree-switcher.nix]: Fair enough. I just used what upstream docs said as I'm not familiar with `fish`. I'll look into implementing this fix on my laptop shortly.

File: pkgs/by-name/gi/git-worktree-switcher/disable-update.patch
Comment (mateusauler) [inline pkgs/by-name/gi/git-worktree-switcher/disable-update.patch:78]:
```suggestion
 	directory=$(git worktree list --porcelain 2> /dev/null | sed -n '/'"${arg:-.}"'/{s/^worktree\s*//p;q}')
```

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (mateusauler) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:13]:
```suggestion
  version = "0.2.4";
```

File: pkgs/by-name/gi/git-worktree-switcher/package.nix
Comment (mateusauler) [inline pkgs/by-name/gi/git-worktree-switcher/package.nix:19]:
```suggestion
    hash = "sha256-N+bDsLEUM6FWhyliUav2n5hwMa5EEuVPoIK+Cja0DxA=";
```

**Review by mateusauler — COMMENTED** [review]
@jiriks74 FYI, I just released v0.2.4.

**Comment by jiriks74** [comment]
Thanks for the notice @mateusauler.

I've simplified the patch to removing only the update calls, not the functions themselves (less potential breakages in the future) and updated the package to `0.2.4`.

**Review by xanderio — APPROVED** [review]

**Review by leona-ya — APPROVED** [review]

**Comment by jiriks74** [comment]
@mateusauler You can make a pr and add yourself as a mainteiner now.

**Pull request merged**
